lib/devise/parameter_sanitizer.rb

```diff
@@ -1,99 +1,179 @@
+      account_update: [:password, :password_confirmation, :current_password]
+    }
 
     def initialize(resource_class, resource_name, params)
-      @resource_class = resource_class
-      @resource_name  = resource_name
+      @auth_keys      = extract_auth_keys(resource_class)
       @params         = params
-      @blocks         = Hash.new
+      @resource_name  = resource_name
+      @permitted      = {}
+
+      DEFAULT_PERMITTED_ATTRIBUTES.each_pair do |operation, keys|
+        permit(operation, keys: keys)
+      end
     end
 
-    def for(kind, &block)
-      if block_given?
-        @blocks[kind] = block
+    # Sanitize the parameters for a specific +operation+.
+    #
+    # === Arguments
+    #
+    # * +operation+ - A +Symbol+ with the operation that the controller is
+    #   performing, like +sign_up+, +sign_in+, etc.
+    #
+    # === Examples
+    #
+    #    # Inside the `RegistrationsController#create` action.
+    #    resource = build_resource(devise_parameter_sanitizer.sanitize(:sign_up))
+    #    resource.save
+    #
+    # Returns an +ActiveSupport::HashWithIndifferentAccess+ with the permitted
+    # attributes.
+    def sanitize(operation)
+      permissions = @permitted[operation]
+
+      if respond_to?(operation, true)
+        deprecate_instance_method_sanitization(operation)
+        return send(operation)
+      end
+
+      if permissions.respond_to?(:call)
+        cast_to_hash permissions.call(default_params)
+      elsif permissions.present?
+        cast_to_hash permit_keys(default_params, permissions)
       else
-        default_for(kind)
+        unknown_operation!(operation)
       end
     end
 
-    def sanitize(kind)
-      if block = @blocks[kind]
-        block.call(default_params)
+    def permit(operation, keys: nil, except: nil, &block)
+      if block_given?
+        @permitted[operation] = block
+      end
+
+      if keys.present?
+        @permitted[operation] ||= @auth_keys.dup
+        @permitted[operation].concat(keys)
+      end
+
+      if except.present?
+        @permitted[operation] ||= @auth_keys.dup
+        @permitted[operation] = @permitted[operation] - except
+      end
+    end
+
+    def for(operation, &block)
+      if block_given?
+        deprecate_for_with_block(operation)
+        permit(operation, &block)
       else
-        default_sanitize(kind)
+        deprecate_for_without_block(operation)
+        @permitted[operation] or unknown_operation!(operation)
       end
     end
 
     private
 
-    def default_for(kind)
-      raise ArgumentError, "a block is expected in Devise base sanitizer"
-    end
-
-    def default_sanitize(kind)
-      default_params
+    # Cast a sanitized +ActionController::Parameters+ to a +HashWithIndifferentAccess+
+    # that can be used elsewhere.
+    #
+    # Returns an +ActiveSupport::HashWithIndifferentAccess+.
+    def cast_to_hash(params)
+      if params
+        params.to_hash.with_indifferent_access
```

Comment:
to_hash returns all parameters not just the permitted ones - to_h returns a HWIA with only the allowed parameters.